File: browser_handler.py
from playwright.async_api import async_playwright
import logging
from typing import Dict

logger = logging.getLogger(__name__)

class BrowserHandler:
    """Handler for headless browser operations using Playwright"""
    
    def __init__(self):
        self.playwright = None
        self.browser = None
        
    async def __aenter__(self):
        self.playwright = await async_playwright().start()
        self.browser = await self.playwright.chromium.launch(headless=True)
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.browser:
            await self.browser.close()
        if self.playwright:
            await self.playwright.stop()
    
    async def fetch_quiz_page(self, url: str) -> Dict[str, str]:
        """
        Fetch and render a quiz page, extracting all content
        """
        try:
            page = await self.browser.new_page()
            
            # Navigate with extended timeout and wait for network idle
            await page.goto(url, wait_until='networkidle', timeout=30000)
            
            # Wait a bit more for any delayed JavaScript execution
            await page.wait_for_timeout(2000)
            
            # Extract the full HTML content
            html_content = await page.content()
            
            # Extract the visible text content
            text_content = await page.evaluate("""
                () => {
                    return document.body.innerText;
                }
            """)
            
            # Try to extract specific elements that commonly contain quiz content
            result_content = ""
            try:
                result_element = await page.query_selector("#result")
                if result_element:
                    result_content = await result_element.inner_text()
            except:
                pass
            
            await page.close()
            
            logger.info(f"Successfully fetched quiz page from {url}")
            logger.debug(f"Text content length: {len(text_content)}")
            
            return {
                "html": html_content,
                "text": text_content,
                "result": result_content if result_content else text_content,
                "url": url
            }
            
        except Exception as e:
            logger.error(f"Error fetching quiz page from {url}: {e}")
            raise
    
    async def download_file(self, url: str, output_path: str) -> str:
        """
        Download a file using the browser
        """
        try:
            page = await self.browser.new_page()
            
            # Navigate to the file URL
            response = await page.goto(url, wait_until='networkidle', timeout=30000)
            
            if response.status != 200:
                raise Exception(f"Failed to download file: HTTP {response.status}")
            
            # Get the content
            content = await response.body()
            
            # Save to file
            with open(output_path, 'wb') as f:
                f.write(content)
            
            await page.close()
            
            logger.info(f"Successfully downloaded file from {url} to {output_path}")
            return output_path
            
        except Exception as e:
            logger.error(f"Error downloading file from {url}: {e}")
            raise

File: browser-handler.js
const puppeteer = require('puppeteer');
const axios = require('axios');

let browser = null;

async function getBrowser() {
    if (!browser) {
        console.log('[BROWSER] Launching Puppeteer...');
        browser = await puppeteer.launch({
            headless: 'new',
            args: [
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage',
                '--disable-gpu',
                '--no-first-run',
                '--no-zygote',
                '--single-process'
            ]
        });
        console.log('[BROWSER] Puppeteer launched successfully');
    }
    return browser;
}

async function extractQuizFromPage(url) {
    const br = await getBrowser();
    const page = await br.newPage();
    
    try {
        console.log(`[BROWSER] Navigating to: ${url}`);
        
        await page.goto(url, {
            waitUntil: 'networkidle2',
            timeout: 30000
        });

        // Wait for page to render
        await page.waitForTimeout(2000);

        console.log('[BROWSER] Page loaded, extracting content...');

        // Extract all text content
        const pageData = await page.evaluate(() => {
            const getText = (element) => {
                const text = element.innerText || element.textContent || '';
                return text.trim();
            };

            // Get the main question
            let question = '';
            const questionElements = document.querySelectorAll('h1, h2, h3, .question, [class*="question"]');
            if (questionElements.length > 0) {
                question = getText(questionElements[0]);
            } else {
                // Fallback: get first heading or prominent text
                const body = document.body;
                question = getText(body).split('\n')[0];
            }

            // Get all visible text content
            const allText = getText(document.body);

            // Find all links (potential file downloads)
            const links = Array.from(document.querySelectorAll('a')).map(a => ({
                url: a.href,
                text: getText(a)
            })).filter(link => link.url && link.url.startsWith('http'));

            // Find submit URL (look for specific patterns)
            let submitUrl = null;
            const submitPatterns = ['/submit', 'submit', 'answer'];
            
            // Check in text content
            const urlRegex = /https?:\/\/[^\s<>"]+/g;
            const foundUrls = allText.match(urlRegex) || [];
            
            for (const foundUrl of foundUrls) {
                for (const pattern of submitPatterns) {
                    if (foundUrl.includes(pattern)) {
                        submitUrl = foundUrl;
                        break;
                    }
                }
                if (submitUrl) break;
            }

            return {
                question,
                content: allText,
                links,
                submitUrl
            };
        });

        console.log('[BROWSER] Extracted data from page');
        
        // Filter links to find file downloads
        const fileLinks = pageData.links.filter(link => {
            const lower = (link.url + ' ' + link.text).toLowerCase();
            return lower.includes('download') || 
                   lower.includes('.csv') || 
                   lower.includes('.pdf') ||
                   lower.includes('.xlsx') ||
                   lower.includes('.xls') ||
                   lower.includes('.json') ||
                   lower.includes('.jpg') ||
                   lower.includes('.png') ||
                   lower.includes('.mp3') ||
                   lower.includes('.wav') ||
                   lower.includes('.mp4') ||
                   lower.includes('file');
        });

        return {
            question: pageData.question,
            content: pageData.content,
            submitUrl: pageData.submitUrl,
            files: fileLinks
        };

    } catch (error) {
        console.error('[BROWSER] Error extracting quiz:', error.message);
        throw error;
    } finally {
        await page.close();
    }
}

async function downloadFile(url) {
    console.log(`[DOWNLOAD] Downloading: ${url}`);
    
    try {
        const response = await axios.get(url, {
            responseType: 'arraybuffer',
            timeout: 30000,
            maxContentLength: 50 * 1024 * 1024, // 50MB
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        });

        console.log(`[DOWNLOAD] Downloaded ${response.data.length} bytes`);
        return Buffer.from(response.data);
    } catch (error) {
        console.error(`[DOWNLOAD] Error downloading file:`, error.message);
        throw error;
    }
}

async function closeBrowser() {
    if (browser) {
        await browser.close();
        browser = null;
        console.log('[BROWSER] Browser closed');
    }
}

// Cleanup on exit
process.on('exit', closeBrowser);
process.on('SIGINT', async () => {
    await closeBrowser();
    process.exit();
});

module.exports = {
    extractQuizFromPage,
    downloadFile,
    closeBrowser
};